haystack_experimental/components/generators/ollama/chat/chat_generator.py

```diff
@@ -0,0 +1,244 @@
+
+    if text_contents:
+        ollama_msg["content"] = text_contents[0]
+    if tool_calls:
+        ollama_tool_calls = []
+        for tc in tool_calls:
+            # Ollama does not support tool call id, so we ignore it
+            ollama_tool_calls.append(
+                {
+                    "type": "function",
+                    "function": {"name": tc.tool_name, "arguments": tc.arguments},
+                }
+            )
+        ollama_msg["tool_calls"] = ollama_tool_calls
+    return ollama_msg
+
+
+@component()
+class OllamaChatGenerator(chatgenerator_base_class):
+    """
+    Supports models running on Ollama.
+
+    Find the full list of supported models [here](https://ollama.ai/library).
+
+    Usage example:
+    ```python
+    from haystack_experimental.components.generators.ollama import OllamaChatGenerator
+    from haystack_experimental.dataclasses import ChatMessage
+
+    generator = OllamaChatGenerator(model="zephyr",
+                                url = "http://localhost:11434",
+                                generation_kwargs={
+                                "num_predict": 100,
+                                "temperature": 0.9,
+                                })
+
+    messages = [ChatMessage.from_system("\nYou are a helpful, respectful and honest assistant"),
+    ChatMessage.from_user("What's Natural Language Processing?")]
+
+    print(generator.run(messages=messages))
+    ```
+    """
+
+    def __init__(
+        self,
+        model: str = "orca-mini",
+        url: str = "http://localhost:11434",
+        generation_kwargs: Optional[Dict[str, Any]] = None,
+        timeout: int = 120,
+        streaming_callback: Optional[Callable[[StreamingChunk], None]] = None,
+        tools: Optional[List[Tool]] = None,
+    ):
+        """
+        Creates an instance of OllamaChatGenerator.
+
+        :param model:
+            The name of the model to use. The model should be available in the running Ollama instance.
+        :param url:
+            The URL of a running Ollama instance.
+        :param generation_kwargs:
+            Optional arguments to pass to the Ollama generation endpoint, such as temperature,
+            top_p, and others. See the available arguments in
+            [Ollama docs](https://github.com/ollama/ollama/blob/main/docs/modelfile.md#valid-parameters-and-values).
+        :param timeout:
+            The number of seconds before throwing a timeout error from the Ollama API.
+        :param streaming_callback:
+            A callback function that is called when a new token is received from the stream.
+            The callback function accepts StreamingChunk as an argument.
+        :param tools:
+            A list of tools for which the model can prepare calls.
+            Not all models support tools. For a list of models compatible with tools, see the
+            [models page](https://ollama.com/search?c=tools).
+        """
+        ollama_integration_import.check()
```

Comment:
Isn't this going to fail if we are extending object instead of OllamaChatGenerator

Reply:
Ok gotcha now, I was trying to figure out why we extend object when `not ollama_integration_import.is_successful()` and we basically do that only to enable smooth functioning of this `ollama-haystack` installed or not import mechanism, correct? If so, perhaps one line of comment would help with maintenance here so we don't forget. Perhaps only commenting with the above [comment](https://github.com/deepset-ai/haystack-experimental/pull/106#discussion_r1782772497).

Reply:
I will add a comment explaining this.